db/view.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/db/description"
+)
+
+func (db *db) addView(
+	ctx context.Context,
+	txn datastore.Txn,
+	inputQuery string,
+	sdl string,
+) ([]client.CollectionDefinition, error) {
+	// Wrap the given query as part of the GQL query object - this simplifies the syntax for users
+	// and ensures that we can't be given mutations.  In the future this line should disappear along
+	// with the all calls to the parser appart from `ParseSDL` when we implement the DQL stuff.
+	query := fmt.Sprintf(`query { %s }`, inputQuery)
+
+	newDefinitions, err := db.parser.ParseSDL(ctx, sdl)
+	if err != nil {
+		return nil, err
+	}
+
+	ast, err := db.parser.BuildRequestAST(query)
+	if err != nil {
+		return nil, err
+	}
+
+	req, errs := db.parser.Parse(ast)
+	if len(errs) > 0 {
+		return nil, errors.Join(errs...)
+	}
+
+	if len(req.Queries) == 0 || len(req.Queries[0].Selections) == 0 {
+		return nil, NewErrInvalidViewQuery(inputQuery)
+	}
+
+	baseQuery, ok := req.Queries[0].Selections[0].(*request.Select)
+	if !ok {
+		return nil, NewErrInvalidViewQuery(inputQuery)
+	}
+
+	for i, definition := range newDefinitions {
+		definition.Description.BaseQuery = baseQuery
+		newDefinitions[i] = definition
+	}
+
+	existingCollections, err := db.getAllCollections(ctx, txn)
+	if err != nil {
+		return nil, err
+	}
+
+	existingDefinitions := make([]client.CollectionDefinition, len(existingCollections))
+	for i := range existingCollections {
+		existingDefinitions[i] = existingCollections[i].Definition()
+	}
+
+	err = db.parser.SetSchema(ctx, txn, append(existingDefinitions, newDefinitions...))
```

Comment:
thought: It would be much nicer to have a `AddSchema` method on parser instead of having to query existing schemas to add a new one.

Reply:
Food for thought, but I'm not sure it is worth the hassle that would mean in the parser/gql code.  That hassle would also increase the cost of adding any new query languages who would need to re-write their own flavour of that hassle in order to satisfy the interface.  Performance isn't a concern in these functions.

Reply:
At the moment it's a hassle that is needed every time there is a call to `SetSchema`. Don't you think it would be better to have that in just one location instead?

Reply:
Not every time, the existing are required when updating schema (and in the future, views).  I also see replacing everything with a new, clean instance as safer than mutating them.

Reply:
> Don't you think it would be better to have that in just one location instead?

We can do this without hosting it inside the Parser interface.